rust/experimental/query_engine/expressions/src/scalars/text_scalar_expression.rs

```diff
@@ -165,6 +179,123 @@ impl Expression for ReplaceTextScalarExpression {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct MatchTextScalarExpression {
+    query_location: QueryLocation,
+    input_expression: Box<ScalarExpression>,
+    pattern: Box<ScalarExpression>,
+}
+
+impl MatchTextScalarExpression {
+    pub fn new(
+        query_location: QueryLocation,
+        input_expression: ScalarExpression,
+        pattern: ScalarExpression,
+    ) -> MatchTextScalarExpression {
+        Self {
+            query_location,
+            input_expression: input_expression.into(),
+            pattern: pattern.into(),
+        }
+    }
+
+    pub fn get_input_expression(&self) -> &ScalarExpression {
+        &self.input_expression
+    }
+
+    pub fn get_pattern(&self) -> &ScalarExpression {
+        &self.pattern
+    }
+
+    pub(crate) fn try_resolve_value_type(
+        &self,
+        _pipeline: &PipelineExpression,
+    ) -> Result<Option<ValueType>, ExpressionError> {
+        // Match always returns a boolean value (true if pattern matches, false otherwise)
+        Ok(Some(ValueType::Boolean))
+    }
+
+    pub(crate) fn try_resolve_static(
+        &self,
+        _pipeline: &PipelineExpression,
+    ) -> Result<Option<ResolvedStaticScalarExpression<'_>>, ExpressionError> {
+        // Match cannot be resolved statically since it depends on runtime input
+        Ok(None)
+    }
+}
+
+impl Expression for MatchTextScalarExpression {
+    fn get_query_location(&self) -> &QueryLocation {
+        &self.query_location
+    }
+
+    fn get_name(&self) -> &'static str {
+        "MatchTextScalarExpression"
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ExtractTextScalarExpression {
+    query_location: QueryLocation,
+    input_expression: Box<ScalarExpression>,
+    pattern_expression: Box<ScalarExpression>,
+    capture_name_or_index: Box<ScalarExpression>,
+}
+
+impl ExtractTextScalarExpression {
+    pub fn new(
+        query_location: QueryLocation,
+        input_expression: ScalarExpression,
+        pattern_expression: ScalarExpression,
+        capture_name_or_index: ScalarExpression,
+    ) -> ExtractTextScalarExpression {
+        Self {
+            query_location,
+            input_expression: input_expression.into(),
+            pattern_expression: pattern_expression.into(),
+            capture_name_or_index: capture_name_or_index.into(),
+        }
+    }
+
+    pub fn get_input_expression(&self) -> &ScalarExpression {
+        &self.input_expression
+    }
+
+    pub fn get_pattern_expression(&self) -> &ScalarExpression {
+        &self.pattern_expression
+    }
+
+    pub fn get_capture_name_or_index(&self) -> &ScalarExpression {
+        &self.capture_name_or_index
+    }
+
+    pub(crate) fn try_resolve_value_type(
+        &self,
+        _pipeline: &PipelineExpression,
+    ) -> Result<Option<ValueType>, ExpressionError> {
+        // Extract always returns a string value (the extracted text)
```

Comment:
What happens if the inputs are invalid or it can't find what it is looking for? Do we return empty string?